Zend/zend_vm_def.h

```diff
@@ -6405,6 +6405,69 @@ ZEND_VM_C_LABEL(isset_no_object):
 	ZEND_VM_NEXT_OPCODE_CHECK_EXCEPTION();
 }
 
+ZEND_VM_HANDLER(199, ZEND_ARRAY_KEY_EXISTS, CV|TMPVAR|CONST, CV|TMPVAR|CONST)
+{
+	USE_OPLINE
+
+	zend_free_op free_op1, free_op2;
+	zval *key, *subject;
+	HashTable* ht;
+	int result;
+
+	SAVE_OPLINE();
+
+	key = GET_OP1_ZVAL_PTR_UNDEF(BP_VAR_R);
+	subject = GET_OP2_ZVAL_PTR_UNDEF(BP_VAR_R);
+
+ZEND_VM_C_LABEL(try_again):
+
+	if (UNEXPECTED(Z_ISREF_P(subject)) || UNEXPECTED(Z_ISREF_P(key))) {
```

Comment:
This isn't quite what I had in mind regarding the reference handling. The idea is to do something like this:

```x
ZEND_VM_C_LABEL(try_again_subject):
	if (EXPECTED(Z_TYPE_P(subject) == IS_ARRAY)) {
		ht = Z_ARRVAL_P(subject);
	} else if (UNEXPECTED(Z_TYPE_P(subject) == IS_OBJECT)) {
		ht = Z_OBJPROP_P(subject);
	} else if (Z_TYPE_P(Z_ISREF_P(subject)) {
        subject = Z_REFVAL_P(subject);
        ZEND_VM_C_GOTO(try_again_subject);
    } else {
		if (OP1_TYPE == IS_CV && UNEXPECTED(Z_TYPE_INFO_P(key) == IS_UNDEF)) {
			key = GET_OP1_UNDEF_CV(key, BP_VAR_R);
		}
		if (OP2_TYPE == IS_CV && UNEXPECTED(Z_TYPE_INFO_P(subject) == IS_UNDEF)) {
			subject = GET_OP2_UNDEF_CV(subject, BP_VAR_R);
		}
		zend_internal_type_error(EX_USES_STRICT_TYPES(), "array_key_exists() expects parameter 2 to be array, %s given", zend_get_type_by_const(Z_TYPE_P(subject)));
		FREE_OP2();
		FREE_OP1();
		ZEND_VM_SMART_BRANCH(result, 0);
		ZVAL_NULL(EX_VAR(opline->result.var));
		ZEND_VM_NEXT_OPCODE_CHECK_EXCEPTION();
    }
```

The point is that the check for the reference is only done after the most likely case (subject is an array) is already handled, so that rare cases (references) don't have a runtime impact on common ones (array).

Reply:
Hah, got it!